### ISSUE TITLE
Possibly a bug in quasi-quotations

### ISSUE THREAD

**Issue by shayan-najd** [idris-lang/Idris-dev#2490]
The following does not work:
~~~
test : TT
test = let f : TT = `( \ y : Char => y)
       in  `( ~f 'c' )
~~~
Idris (0.9.18.1-git:fa0061d) REPL returns the following error message:

> When checking right hand side of test:
>      unquotation does not have a function type (_)

**Comment by david-christiansen** [comment]
The error message is definitely bad, but I don't know that we can do much better without additional annotations, at least not without restructuring the elaborator for this. It might be possible to assume non-dependent types for the function position and attempt to synthesize a type annotation, but I'm afraid that this might have other bad effects.

How would you feel about the ability to place type annotations on antiquotations, kind of like the annotations that full quotations can accept?

**Comment by shayan-najd** [comment]
I have difficulty understanding what (and why) type information is missing in the first place. Can you explain further?

I have taken a look at the related code (in elaboration). There you write the following comment:

~~~
-- First extract the unquoted subterms, replacing them with fresh
-- names in the quasiquoted term. Claim their reflections to be
-- an inferred type (to support polytypic quasiquotes).
~~~

(I am not sure if I understand what you mean by polytypic quasiquotes above.) 
Can't you just infer the (or, access the inferred) type of the unquoted terms and claim the type of the corresponding fresh variables to be of the inferred types?

It implies that every time you splice something you should have access to the type of its reflected term.

**Comment by david-christiansen** [comment]
The problem here is that we can't infer the type of the function position. It's not sufficient to infer something like `Char -> Char` - what if the type had been `(c : Char) -> if c == 'c' then Char else Nat`? In general, the elaborator will always need to know what the type of the things being applied are, which it normally gets though looking them up in the global context. When you have an unconstrained hole, it can't do this.  Additionally, there may have been implicits to solve.

> It implies that every time you splice something you should have access to the type of its reflected term.

In argument positions, this is often inferrable.

**Comment by shayan-najd** [comment]
Thanks for the response. 

> It's not sufficient to infer something like Char -> Char - what if the type had been 
> `(c : Char) -> if c == 'c' then Char else Nat?`

I do not see how type of  ` \ y : Char => y` can be anything other than `Char -> Char`, but I guess I understand your concern: in general we cannot fully infer type of terms (no surprise there). 

But, why should it behave differently than the following?
~~~
test : TT
test = `((\ y : Char => y) 'c')
~~~
Precisely, what information is added by manual inlining? 

I expect quotations of "later stages" to be type-checked/inferred only when their anti-quotes are fully spliced. Such splicing does not actually need to happen in practice, for the performance reasons, but type-checking/inference should behave as if they would. It all boils down to proper propagation of meta-variables (see the following discussion).

> In general, the elaborator will always need to know what the type of the things being applied are, 
> which it normally gets though looking them up in the global context.

Does it mean type of "the things being applied" should not include meta-variables? 

Consider the following:
~~~
test2 : TT
test2 = `( ( \ y => y) 'c' )
~~~
Above, I expect type of `(\ y => y)`, at some intermediate step of typechecking / inference, to be something like `a0 -> a0`, where `a0` is a fresh meta-variable. Then, in the later steps, `a0` should be set equal to `Char` (or whatever type of a character literal is). Similarly, if `\ y => y` was written in a quotation, its internally inferred type should be something like `a0 -> a0`, and the meta-variable 'a0' should be refreshed *each* time it is spliced! 

For instance, the following should type-check as-is:

~~~
test3 : TT
test3 =  let f : TT = `( \ y => y)
             in  `( ~f 'c' , ~f True)
~~~

**Comment by david-christiansen** [comment]
> I expect quotations of "later stages" to be type-checked/inferred only when their anti-quotes are fully spliced. 

I think we've found the root here - that expectation is not going to be fulfilled. We can't necessarily know what will be spliced in, so each quotation is checked alone. An alternative might be some kind of partial evaluation (like I think you're proposing), but that would be very complicated and, I would guess, lead to difficult-to-predict errors. The current model is straightforward (quotations are type checked on their own), but it does look like we need to have type annotations for antiquotations.

How would you propose handling something like:
```
main : IO ()
main =
  do tm <- getLine
     let Right tm' = parseQuotedTerm tm
                | Left err => putStrLn err
     printLn `(~tm' 17)         
```

The system has no way to know what the contents of `tm'` are when type checking the quotation!

**Comment by shayan-najd** [comment]
> We can't necessarily know what will be spliced in, so each quotation is checked alone. 

As you mentioned, there are models (e.g. Template Haskell) in which one always knows what will be spliced in. Though, such models are sometimes restrictive (besides being complicated).  

If we follow the implemented model, I guess type annotations should be provided for the quotations rather than the anti-quotations.

**Comment by david-christiansen** [comment]
Type annotations can already be provided for the whole quotation. That's not enough, though - your example shows that we sometimes need more type information in order to elaborate the quotation.

**Comment by shayan-najd** [comment]
> Type annotations can already be provided for the whole quotation.

No, quotation mechanism for TT loses type information:

~~~
test : TT
test = let f : TT = `( \ y : Char => y : Char -> Char)
          in  `( ~f 'c' )
~~~

**Comment by david-christiansen** [comment]
Right, the type doesn't stick around once the quotation is created. It's used when elaborating the quotation, however. For example, `` `(() : Type) `` yields a different quoted term than `` `(() : ()) ``.

**Comment by david-christiansen** [comment]
I still don't think it's at all unreasonable or unexpected to think that each quotation should quote a well-typed term, regardless of the surrounding code. Certainly not where we have quotations that don't track types!

Your issue could be solved if we supported a syntax like

    foo : TT -> TT
    foo f = `( ~(f : Char -> Char) 'c') 

where the annotation on the antiquotation gives enough info to elaborate the term.

**Comment by shayan-najd** [comment]
> Right, the type doesn't stick around once the quotation is created.

I guess this behaviour is OK for TT, but I expect types to stick around for the "typed quasi-quotes". 
Would they?

Is it correct to assume a quasi-quotation
~~~
foo : A
foo = `(... ~(M) ...) 
~~~
first elaborates to the following quotation
~~~
foo : A
foo = let x0 : A = M in `(... x0 ...)
-- where x0 is a fresh name
~~~
and then quotations are elaborated to the code representing the AST they denote?
